unity-renderer/Assets/DCLServices/QuestsService/QuestsService.cs

```diff
@@ -0,0 +1,138 @@
+        internal CancellationTokenSource userSubscribeCt = null;
+
+        public QuestsService(IClientQuestsService clientQuestsService)
+        {
+            this.clientQuestsService = clientQuestsService;
+        }
+
+        public void SetUserId(string userId)
+        {
+            if(userId == this.userId)
+                return;
+
+            this.userId = userId;
+            // Definitions are not user specific, so we only need to clear the state cache
+            stateCache.Clear();
+            userSubscribeCt?.Cancel();
+            userSubscribeCt?.Dispose();
+            userSubscribeCt = null;
+
+            if (!string.IsNullOrEmpty(userId))
+            {
+                userSubscribeCt = new CancellationTokenSource();
+                Subscribe(userSubscribeCt.Token).Forget();
+            }
+        }
+
+        private async UniTaskVoid Subscribe(CancellationToken ct)
+        {
+            var  enumerator = clientQuestsService.Subscribe(new UserAddress { UserAddress_ = userId }).GetAsyncEnumerator(ct);
+            try
+            {
+                while (await enumerator.MoveNextAsync())
+                {
+                    var userUpdate = enumerator.Current;
+
+                    switch (userUpdate.MessageCase)
+                    {
+                        case UserUpdate.MessageOneofCase.QuestStateUpdate:
+                            stateCache[userUpdate.QuestStateUpdate.QuestData.QuestInstanceId] = userUpdate.QuestStateUpdate.QuestData;
+                            OnQuestUpdated?.Invoke(userUpdate.QuestStateUpdate.QuestData);
+                            break;
+                        case UserUpdate.MessageOneofCase.NewQuestStarted:
+                            stateCache[userUpdate.NewQuestStarted.QuestInstanceId] = userUpdate.NewQuestStarted;
+                            OnQuestStarted?.Invoke(userUpdate.NewQuestStarted);
+                            break;
+                    }
+                }
+            }
+            finally
+            {
+                await enumerator.DisposeAsync();
+            }
+        }
+
+        public async UniTask<StartQuestResponse> StartQuest(string questId)
+        {
+            if(string.IsNullOrEmpty(userId))
+                throw new UserIdNotSetException();
+
+            return await clientQuestsService.StartQuest(new StartQuestRequest { QuestId = questId, UserAddress = userId });
+        }
+
+        public async UniTask<AbortQuestResponse> AbortQuest(string questInstanceId)
+        {
+            if(string.IsNullOrEmpty(userId))
+                throw new UserIdNotSetException();
+
+            return await clientQuestsService.AbortQuest(new AbortQuestRequest { QuestInstanceId = questInstanceId, UserAddress = userId });
+        }
+
+        public async UniTask<Quest> GetDefinition(string questId, CancellationToken cancellationToken = default)
```

Comment:
Not a big deal but you can return a UniTask without `async` so it will be awaited in the outer layer